presto-native-execution/presto_cpp/main/types/ExpressionOptimizer.cpp

```diff
@@ -0,0 +1,150 @@
+  velox::exec::ExprSet exprSet({expr}, &execCtx);
+  velox::exec::EvalCtx evalCtx(&execCtx, &exprSet, data.get());
+
+  const velox::SelectivityVector singleRow(1);
+  std::vector<velox::VectorPtr> results(1);
+  exprSet.eval(singleRow, evalCtx, results);
+  return results.at(0);
+}
+
+protocol::RowExpressionOptimizationResult getRowExpressionOptimizationResult(
+    const RowExpressionPtr& rowExpression,
+    const std::string& optimizerLevel,
+    const VeloxExprConverter& veloxExprConverter,
+    const expression::VeloxToPrestoExprConverter& veloxToPrestoExprConverter,
+    velox::core::QueryCtx* queryCtx,
+    velox::memory::MemoryPool* pool) {
+  protocol::RowExpressionOptimizationResult result;
+  const auto expr = veloxExprConverter.toVeloxExpr(rowExpression);
+  auto optimized =
+      velox::expression::optimize(expr, queryCtx, pool, kMakeFailExpr);
+
+  if (optimizerLevel == kEvaluated) {
+    try {
+      const auto evalResult = tryEvaluateToConstant(optimized, queryCtx, pool);
+      optimized = std::make_shared<velox::core::ConstantTypedExpr>(evalResult);
+    } catch (const velox::VeloxException& e) {
+      result.expressionFailureInfo =
+          toNativeSidecarFailureInfo(translateToPrestoException(e));
+      result.optimizedExpression = nullptr;
+      return result;
+    } catch (const std::exception& e) {
+      result.expressionFailureInfo =
+          toNativeSidecarFailureInfo(translateToPrestoException(e));
+      result.optimizedExpression = nullptr;
+      return result;
+    }
+  }
+
+  result.optimizedExpression =
+      veloxToPrestoExprConverter.getRowExpression(optimized, rowExpression);
+  return result;
+}
+
+} // namespace
+
+void optimizeExpressions(
+    const proxygen::HTTPHeaders& httpHeaders,
+    const json::array_t& inputRowExpressions,
+    proxygen::ResponseHandler* downstream,
+    folly::Executor* driverExecutor,
+    velox::memory::MemoryPool* pool) {
+  const auto& optimizerLevel =
+      httpHeaders.getSingleOrEmpty(kOptimizerLevelHeader);
+  VELOX_USER_CHECK(
+      (optimizerLevel == kOptimized) || (optimizerLevel == kEvaluated),
+      "Optimizer level should be OPTIMIZED or EVALUATED, received {}.",
+      optimizerLevel);
+  const auto& timezone = httpHeaders.getSingleOrEmpty(kTimezoneHeader);
+  std::unordered_map<std::string, std::string> config(
+      {{velox::core::QueryConfig::kSessionTimezone, timezone},
+       {velox::core::QueryConfig::kAdjustTimestampToTimezone, "true"}});
+  auto queryConfig = velox::core::QueryConfig{std::move(config)};
+  auto queryCtx =
+      velox::core::QueryCtx::create(driverExecutor, std::move(queryConfig));
+
+  TypeParser typeParser;
+  const VeloxExprConverter veloxExprConverter(pool, &typeParser);
+  const expression::VeloxToPrestoExprConverter veloxToPrestoExprConverter(pool);
+  const auto numExpr = inputRowExpressions.size();
+  json j;
+  json::array_t result;
+  for (auto idx = 0; idx < numExpr; idx++) {
```

Comment:
Nit : Use shorter variable names that don't need the structure name in it. The index variable can be i, the inputRowExpr input and rowExpressionOptimizationResult be optimized